### ISSUE TITLE
Undetected overflow in fxarithmetic-shift-left

### ISSUE THREAD

**Issue by weinholt** [cisco/ChezScheme#165]
The `fxarithmetic-shift-left` procedure does not detect overflow under some conditions. Here are some comparisons showing that the upper bits of the result are discarded:

```Scheme
Chez Scheme Version 9.4.1
Copyright 1984-2017 Cisco Systems, Inc.

> (fixnum-width)
61
> (number->string (fxarithmetic-shift-left #x12 58) 16)
"800000000000000"
> (number->string (bitwise-arithmetic-shift-left #x12 58) 16)
"4800000000000000"
> (number->string (fxarithmetic-shift-left #xB00000000000000 2) 16)
"C00000000000000"
> (number->string (bitwise-arithmetic-shift-left #xB00000000000000 2) 16)
"2C00000000000000"
> (number->string (fxarithmetic-shift-left #x-500000000000000 3) 16)
"-800000000000000"
> (number->string (bitwise-arithmetic-shift-left #x-500000000000000 3) 16)
"-2800000000000000"
```

Tested with master (v9.4-132-g8fbe1b1) on a Debian GNU/Linux 8.7 amd64 system.

**Comment by burgerrg** [comment]
Nice find! fxarithmetic-shift-left also gives the wrong error who (fxsll) when the arguments are not fixnums. The logic in library.ss for fxsll and fxarithmetic-shift-left needs to corrected.

**Comment by burgerrg** [comment]
@dybvig, the documentation of fxsll and fxarithmetic-shift-left aren't in sync. fxsll allows fixnum-width shifts, whereas fxarithmetic-shift-left doesn't. fxarithmetic-shift-left doesn't mention an exception on overflow. What would you like me to change?

**Comment by dybvig** [comment]
If you change the shift-count restriction for either, fxsll and fxsrl should behave like fxarithmetic-shift-left and fxarithmetic-shift-right.  The latter are constrained not to allow fixnum-width shifts by R6RS.  fxsll and fxsrl historically allowed fixnum-width shifts, and we didn't change that when adding the R6RS procedures.

The requirement to produce &implementation-restriction on fixnum overflow is mentioned in the introduction of TSPL Section 6.5 and also in the introduction of the R6RS library document Section 11.2.

**Comment by burgerrg** [comment]
Thanks, I didn't change the limits for the shift operators but fixed the overflow detection. Fixed in commit f286688b64f877248ab8d8f00528d19363c6ceba
